src/placeos-core/module_manager.cr

```diff
@@ -275,37 +275,54 @@ module PlaceOS::Core
     # Clustering
     ###############################################################################################
 
-    def on_managed_edge?(mod : Model::Module)
-      mod.on_edge? && own_node?(mod.edge_id.as(String))
-    end
-
-    private getter semaphore = Atomic(Int32).new(0)
+    private getter queued_stabilization_events = Atomic(Int32).new(0)
     private getter stabilize_lock = Mutex.new
+    private STABILIZE_BATCH_SIZE = 32
 
     # Run through modules and load to a stable state.
     #
     # Uses a semaphore to ensure intermediary cluster events don't trigger stabilization.
     def stabilize(nodes : Array(HoundDog::Service::Node)) : Bool
-      semaphore.add(1)
+      queued_stabilization_events.add(1)
       stabilize_lock.synchronize do
-        semaphore.add(-1)
-        return false unless semaphore.get.zero?
+        queued_stabilization_events.add(-1)
```

Comment:
May be cleaner to do this with an [`Atomic::Flag`](https://crystal-lang.org/api/0.36.1/Atomic/Flag.html).

Reply:
Idea is to skip intermediate stabilisation events, rather than set a flag.

Reply:
Ah yep - getting Bool state returned to the last call is important, nevermind.